refactor(common): use getElementById in ViewportScroller.scrollToAnchor

aleesaan opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: [28960](https://github.com/angular/angular/issues/28960)


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

trotyl commented:
If this is preferred then https://github.com/angular/angular/pull/29577 should be reverted as `getElementById` accepts the unescaped value.

Personally I do I agree this one is more size-saving and performant. (It's no longer `fix` but `refactor` now)

petebacondarwin commented:
I note that in #29577, there is a code path to handle `name="..."` as well. I don't think that `getElementById()` would handle this, right?

petebacondarwin commented:
I just checked and if only the `name` attribute is provided then `getElementById` returns `null`.

petebacondarwin commented:
But there is a `document.getElementsByName()` so that would be acceptable.

petebacondarwin commented:
@aleesaan - would you like to modify your PR so that it uses both `getElementById()` and `getElementsByName()` to avoid having to parse the `anchor` string and avoids having to wrap in a `try-catch` block?

petebacondarwin commented:
You would also change the commit message to `refactor...`

aleesaan commented:
Hi @petebacondarwin, it's my first time working on a PR here and I'm trying to figure out why the tests are failing, could you perhaps help?  
Also, in #29577 a test for the `ViewportScroller` was added, but I think that can be removed now?

petebacondarwin commented:
Sure. Let me check the tests now...

petebacondarwin commented:
So the first failure to worry about is `//packages/common/test`. (The `//packages/common/test:test_web_chromium` is the same test set but running in Chromium headless rather than node).

```
Test output for //packages/common/test:test:
Randomized with seed 96424
Started
..........................................................................................................................
 
Failures:
1) BrowserViewportScroller escapes invalid characters selectors 
  Message:
    TypeError: this.document.getElementById is not a function
  Stack:
        at <Jasmine>
        at BrowserViewportScroller.scrollToAnchor (packages/common/src/viewport_scroller.ts:114:44)
        at packages/common/test/viewport_scroller_spec.ts:34:16
        at UserContext.<anonymous> (packages/core/testing/src/testing_internal.ts:147:23)
        ... 

309 specs, 1 failure
 ```

Are you able run this test locally? The best way is to do:

```
yarn bazel test //packages/common/test
```

The failure is because the mock `document` object created in the test:

```
documentSpy = jasmine.createSpyObj('document', ['querySelector']);
```

Does not contain the required methods (i.e. `getElementById()` and `getElementsByName()`).

I think we should keep this test but change it so that the document mock object has these methods and then check that they are called later in the test.

petebacondarwin commented:
The other failing tests (I think) are just size regressions:

```
FAIL: Commit undefined uncompressed main-es2015 fell below expected size by 500 bytes or >1% (expected: 246044, actual: 245437).
```

This is actually a positive change - the size is reducing!

We can work on getting those test to pass once the rest of the code is settled and passing.

aleesaan commented:
Thanks for the quick response!  

I've been trying to run the tests but my computer is taking hours just to pull and build all the files, so I couldn't figure out what was failing. I'm now at building all the packages as mentioned [here](https://github.com/angular/angular/blob/master/docs/BAZEL.md#building-angular), and that should be the last step before I can finally run the tests.

Regarding the `ViewportScroller` test, I don't think we need to check the escaping anymore, right? Shall I replace that test with one that checks that both `getElementById()` and `getElementsByName()` are called? Is that what you meant?

petebacondarwin commented:
> Regarding the ViewportScroller test, I don't think we need to check the escaping anymore, right? Shall I replace that test with one that checks that both getElementById() and getElementsByName() are called? Is that what you meant?

Yes, precisely

petebacondarwin commented on packages/common/test/viewport_scroller_spec.ts:
```diff
@@ -23,18 +23,21 @@ import {BrowserViewportScroller, ViewportScroller} from '../src/viewport_scrolle
-      expect(documentSpy.querySelector).toHaveBeenCalledWith(`#specials\\=${escapedInvalids}`);
-      expect(documentSpy.querySelector)
-          .toHaveBeenCalledWith(`[name='specials\\=${escapedInvalids}']`);
+    describe('scrollToAnchor', () => {
```
While we are here, let's add some empty lines between describe and it blocks, to keep it more inline with other files in the repo.

```suggestion

    describe('scrollToAnchor', () => {
```

petebacondarwin replied:
You could also remove the duplicated copyright block above, and in fact we might as well remove the `{ .. }` block around the whole set of tests, since this is from a long time back when we were not using a bundler before loading the files into the browser.

aleesaan replied:
Yeah better! I wasn't sure about the code style so I left it as it was in the file.

petebacondarwin commented on packages/common/test/viewport_scroller_spec.ts:
```diff
@@ -23,18 +23,21 @@ import {BrowserViewportScroller, ViewportScroller} from '../src/viewport_scrolle
-          .toHaveBeenCalledWith(`[name='specials\\=${escapedInvalids}']`);
+    describe('scrollToAnchor', () => {
+      const anchor = 'anchor';
+      beforeEach(() => {
```
```suggestion

      beforeEach(() => {
```

petebacondarwin commented on packages/common/test/viewport_scroller_spec.ts:
```diff
@@ -23,18 +23,21 @@ import {BrowserViewportScroller, ViewportScroller} from '../src/viewport_scrolle
+        });
+        scroller = new BrowserViewportScroller(documentSpy, {scrollTo: 1}, null!);
+      });
+      it('should call getElementById and getElementsByName when an element is not found by its id or its name',
```
```suggestion

      it('should call getElementById and getElementsByName when an element is not found by its id or its name',
```

petebacondarwin commented on packages/common/test/viewport_scroller_spec.ts:
```diff
@@ -23,18 +23,21 @@ import {BrowserViewportScroller, ViewportScroller} from '../src/viewport_scrolle
+           scroller.scrollToAnchor(anchor);
+           expect(documentSpy.getElementById).toHaveBeenCalledWith(anchor);
+           expect(documentSpy.getElementsByName).toHaveBeenCalledWith(anchor);
+         });
```
If you wanted to be super comprehensive, you could add a test that shows that `scroller.scrollToElement()` is called with the return value of either `getElementById()` or `getElementsByName()`.

aleesaan replied:
Pushed! Let me know if that's what you meant.

As a side question, how should I name my single commits? Should they describe the individual commit (e.g. fixing the test, in this last commit) or should I keep repeating the same name, describing the whole PR?

petebacondarwin replied:
Maybe github. Let's give it a few minutes to see if it catches up.
Meanwhile, when making small post-review changes like this to a PR, it is best to create "fixup" commits. Our merge script is then able to squash these commits into a single commit in the merged branch.

aleesaan replied:
Great thanks for the explanation! So can I prefix whatever message with `fixup!`, like `fixup! resolve comments`? Or each commit message still needs to follow the guidelines (as you did [here](https://github.com/angular/angular/pull/38003/commits))?

petebacondarwin replied:
The easiest way is to use the `git` command line to create the commit. For example the following will create a fixup commit which uses the message from the current most recent commit.

```
git commit -a --fixup=HEAD
```

I think the idea with fixups is that their title must match the commit that is being fixed up, but the body can/should be empty.

aleesaan replied:
Thank you so much again, I got it now!

aleesaan replied:
Github finally picked up the commit 🙂

petebacondarwin reviewed (commented):
This is much better! Thanks.
A little tidy up and then we can fix up the size change files.

petebacondarwin reviewed (approved):
Awesome this is great.  Now we just need to do two things:

* fix up the integration test file-size golden files (see below)
* rename the second commit to a `fixup!` commit, so that will also get squashed

There are two integration tests where we need to change the file-size...

```
Verifying sizes against 'angular/goldens/size-tracking/integration-payloads.json'...
... 
SUCCESS: Commit undefined uncompressed polyfills-es2015 did NOT cross size threshold of 500 bytes or >1% (expected: 36938, actual: 36571). 
SUCCESS: Commit undefined uncompressed 5-es2015 did NOT cross size threshold of 500 bytes or >1% (expected: 751, actual: 747). 
FAIL: Commit undefined uncompressed main-es2015 fell below expected size by 500 bytes or >1% (expected: 246044, actual: 245437).
 
If this is a desired change, please update the size limits in file 'goldens/size-tracking/integration-payloads.json'.
```

```
Verifying sizes against 'goldens/size-tracking/integration-payloads.json'...
...
SUCCESS: Commit undefined uncompressed polyfills-es2015 did NOT cross size threshold of 500 bytes or >1% (expected: 36938, actual: 36571). 
SUCCESS: Commit undefined uncompressed 5-es2015 did NOT cross size threshold of 500 bytes or >1% (expected: 751, actual: 747). 
FAIL: Commit undefined uncompressed main-es2015 fell below expected size by 500 bytes or >1% (expected: 246044, actual: 245437).
 
If this is a desired change, please update the size limits in file 'goldens/size-tracking/integration-payloads.json'.
```

It is worth updating the ones I show above (even the passing ones) since their sizes have all dropped.

aleesaan commented:
In the end I reverted the multiple commits and created a single one to have it tidier. I also have to rebase with master and fix conflicts in the `integration-payloads` file, I'll do that now.

aleesaan committed:
refactor(common): use getElementById in ViewportScroller.scrollToAnchor

This commit uses getElementById and getElementsByName when an anchor scroll happens,
to avoid escaping the anchor and wrapping the code in a try/catch block.

Related to #28960

jelbourn reviewed (approved):
LGTM, using global-approvers instead of size-tracking because this is a (very nice) size decrease (vs an increase)

Reviewed-for: global-approvers

alfaproject commented on packages/common/src/viewport_scroller.ts:
```diff
@@ -111,26 +111,15 @@ export class BrowserViewportScroller implements ViewportScroller {
+      if (elSelectedByName) {
+        this.scrollToElement(elSelectedByName);
+        return;
       }
```
You can reduce the bundle size further:
```ts
      const elSelected = this.document.getElementById(anchor) || this.document.getElementsByName(anchor)[0];
      if (elSelected) {
        this.scrollToElement(elSelected);
      }
```

petebacondarwin replied:
Goodl call @alfaproject

aleesaan replied:
Thanks for all the input guys, since it's my first time contributing I was hesitant about making any changes and only stuck to your suggestions, especially cause I'm still struggling running all the builds and tests 😅 but indeed I will try to be more brave (and critical) for the future!

alfaproject commented on packages/common/src/viewport_scroller.ts:
```diff
@@ -111,26 +111,10 @@ export class BrowserViewportScroller implements ViewportScroller {
+      const elSelected = this.document.getElementById(anchor) || this.document.getElementsByName(anchor)[0];
+      if (elSelected) {
+        this.scrollToElement(elSelected);
+        return;
```
You don't need this return

AndrewKushnir commented:
[Presubmit](http://test/OCL:325089736:BASE:325089767:1596659503452:d7152020)

AndrewKushnir commented:
Hi @aleesaan, it looks like this PR has a conflict with the master branch after merging #30630. Could you please rebase it and resolve conflicts? I'm adding the "cleanup" label for now and will remove it once PR is rebased. Thank you.

AndrewKushnir commented:
@aleesaan, one additional comment: this PR also has conflicts with the patch branch (i.e. `10.0.x` branch), so we'd need a second PR that targets `10.0.x` branch (not `master` branch). The conflict is in the `goldens/size-tracking/integration-payloads.json` file. Could you please create a separate PR based on these changes and the current state of the `10.0.x` branch? Please let us know if you need help with it. Thank you.

aleesaan commented:
@AndrewKushnir I rebased this branch! I will create a separate PR based on the `10.0.x` branch, but I'm still not able to run all tests locally, including the ones about file-size. Maybe I could just open the PR without changing them, then look on CI to see the right value, and update accordingly? If you have suggestions I'm all ears!

AndrewKushnir commented:
Hi @aleesaan, thanks for rebasing this PR and resolving conflicts. It looks like there was a problem with CI (caused by the CircleCI), so some jobs failed. I've rebased this PR once again to restart CI.

> I will create a separate PR based on the 10.0.x branch, but I'm still not able to run all tests locally, including the ones about file-size. Maybe I could just open the PR without changing them, then look on CI to see the right value, and update accordingly?

Sounds good 👍

aleesaan commented:
Looks like I had messed up the tests when fixing conflicts 😅 I fixed them and pushed again, hopefully this time it works!

AndrewKushnir commented:
@aleesaan, it looks like some tests are still failing ([for example here](https://app.circleci.com/pipelines/github/angular/angular/19711/workflows/0b8eaae6-b8a8-4188-896b-c8efa5731c60/jobs/782371)). Could you please have a look?

aleesaan commented:
Yep I noticed, I will fix them as soon as I can!

aleesaan committed:
fixup! refactor(common): use getElementById in ViewportScroller.scrollToAnchor
